### PR TITLE
MTDSA-18777 parser-validator improvements

### PR DESCRIPTION
**Pull request by mattstephens-hmrc**

### CODE REVIEW

File: test/v3/controllers/validators/CreateBFLossValidatorFactorySpec.scala [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  )
+
+  private val validRequestBody  = requestBodyJson()
+  private val parsedRequestBody = CreateBFLossRequestBody(TypeOfLoss.`self-employment`, validBusinessId, validTaxYear, 1000)
+
+  class Test extends MockAppConfig {
+    implicit val appConfig: AppConfig = mockAppConfig
+
+    implicit val todaySupplier: TodaySupplier = new TodaySupplier {
+      override def today(): LocalDate = LocalDate.parse("2022-07-11") // , dateTimeFormatter)
```
Comment (SamuelJacobsHMRC) [inline test/v3/controllers/validators/CreateBFLossValidatorFactorySpec.scala:63]: can remove the comment

File: app/v4/controllers/ListLossClaimsController.scala [hunk 2]
```diff
@@ -46,17 +45,16 @@ class ListLossClaimsController @Inject() (val authService: EnrolmentsAuthService
     authorisedAction(nino).async { implicit request =>
       implicit val ctx: RequestContext = RequestContext.from(idGenerator, endpointLogContext)
 
-      val rawData =
-        ListLossClaimsRawData(nino, taxYearClaimedFor = taxYear, typeOfLoss = typeOfLoss, businessId = businessId, typeOfClaim = typeOfClaim)
+      val validator = validatorFactory.validator(nino, taxYear, typeOfLoss = typeOfLoss, businessId = businessId, typeOfClaim = typeOfClaim)
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/v4/controllers/ListLossClaimsController.scala:48]: cos they're all strings, would be easy to get the order mixed up

File: app/api/controllers/validators/resolvers/ResolveTaxYear.scala [hunk 3]
```diff
@@ -39,42 +41,77 @@ trait ResolvingTaxYear extends Resolver[String, TaxYear] {
...
+    allowIncompleteTaxYear: Boolean = true,
+    incompleteTaxYearError: MtdError = RuleTaxYearNotEndedError,
+    maybeMinimumTaxYear: Option[Int] = None,
+    minimumTaxYearError: MtdError = RuleTaxYearNotSupportedError
+)(implicit todaySupplier: TodaySupplier = new TodaySupplier)
+    extends TaxYearResolving {
+
+  def apply(value: String, maybeFormatError: Option[MtdError], errorPath: Option[String]): Validated[Seq[MtdError], TaxYear] = {
+
+    def validateMinimumTaxYear(parsed: TaxYear): Validated[Seq[MtdError], Unit] = {
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply app/api/controllers/validators/resolvers/ResolveTaxYear.scala:85]: nope :)

[the remark this answers is not in the document]
Comment (SamuelJacobsHMRC) [reply app/api/controllers/validators/resolvers/ResolveTaxYear.scala]: oh yes, think i failed to notice it was in the apply 🤦‍♂️

**Review by LouiseWong98 — APPROVED** [review]

**Pull request merged**